### PR TITLE
⚠️ Add watch label to allow multiple manager instances

### LINKED ISSUE #4004: Enable running multiple versions of CAPI controllers in paralell
**User Story**

As an operator I would like to be able to run multiple versions of CAPI controllers to limit the number of management clusters.

**Detailed Description**

The current guideline to change the version of CAPI controllers which reconcile a set of clusters is to move the CRs of the cluster to a new management cluster containing the desired version of CAPI controllers.

This process becomes difficult with a large number and diversity of clusters.

I want to propose that multiple versions of CAPI controllers can run on the same management clusters. The controllers would only reconcile CRs with a certain label. Example:
`cluster.x-k8s.io/watch-label:v0.3.9` -> reconciled by controllers with version `v0.3.9`
`cluster.x-k8s.io/watch-label:v0.3.11` -> reconciled by controllers with version `v0.3.11`

**Anything else you would like to add:**

Using labels for this purpose is not optimal as they are not very structured. They do however make it easy to enable this functionality quickly.

/kind feature

### PR DESCRIPTION
**Pull request by MarcelMue**
**What this PR does / why we need it**:

This PR adds a flag which defines which CRs will be processed depending on a given label.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4004

### CODE REVIEW

File: api/v1alpha4/common_types.go
Comment (MarcelMue) [inline api/v1alpha4/common_types.go:60]: I struggled with naming here - this seemed most consistent with the already existing labels.

File: controllers/types.go
Comment (MarcelMue) [inline controllers/types.go:20]: I am not certain about this. Introducing a Config struct seemed okay in case we want to add more such flags in the future. Currently just passing the string around would also be fine.

Does anyone have any strong opinions here?

File: util/predicates/generic_predicates.go
Comment (MarcelMue) [inline util/predicates/generic_predicates.go:182]: Something similar was implemented before but then removed. I tried to follow that structure and naming.

**Comment by JoelSpeed** [comment]
I think this might need some further discussion before we review. Personally I'd like to understand what the benefit of this approach is over using namespaces to separate the resources that should be reconciled by particular controllers

File: util/predicates/generic_predicates.go
Comment (fabriziopandini) [reply util/predicates/generic_predicates.go:182]: If there are not performance concerns, I would prefer to keep paused and filter by label as a separated predicates

File: api/v1alpha4/common_types.go
Comment (fabriziopandini) [reply api/v1alpha4/common_types.go:60]: I have no brilliant suggestions, but I would like to have something like "watch-filter" or "watch-selector" to make it clear that this value will the used to filter out / to selected

File: controllers/types.go
Comment (fabriziopandini) [reply controllers/types.go:20]: TBD this seems overkill for a single flag...

File: main.go
Comment (fabriziopandini) [inline main.go:62]: might be I'm missing something, but I can't find where this flag gets passed to the reconcilers.

**Review by fabriziopandini — COMMENTED** [review]
/ok-to-test

If this behaviour applies to all the objects managed by Cluster API, we should get this documented in the contract, and implemented also in CAPBK, KCP & CAPD too

**Comment by MarcelMue** [comment]
> I think this might need some further discussion before we review. Personally I'd like to understand what the benefit of this approach is over using namespaces to separate the resources that should be reconciled by particular controllers

There has been discussion on this topic in several different places so far - it was also acknowledged here:
https://github.com/kubernetes-sigs/cluster-api/pull/4074

Google doc which was shared with sig-cluster-lifecycle:
https://docs.google.com/document/d/1Icttu56PmbDm70g8GyXupRq3MX4NADuGDoVhjwQwHMo/

Discussions during sync (just two examples):
https://www.youtube.com/watch?t=649&v=FLUV7QB2QL8
https://www.youtube.com/watch?t=269&v=tV1yoeypfWk

To answer your question briefly:
The namespace separation is not well suited for multi tenancy (in particular with multiple core controllers in different versions) as it implies moving CRs between namespaces. 

Please let me know if there is a clear concern here - I am open to discuss but there has been IMO a lot of discussion leading up to this PR already.

File: main.go
Comment (MarcelMue) [reply main.go]: This was on me - reviewed my code 3 times and missed the obvious :sweat:

File: api/v1alpha4/common_types.go
Comment (MarcelMue) [reply api/v1alpha4/common_types.go]: I changed it to `watch-filter` and follows that naming - WDYT?

File: controllers/types.go
Comment (MarcelMue) [reply controllers/types.go]: I changed it to string for now, still fine to do either.

**Comment by MarcelMue** [comment]
> If this behaviour applies to all the objects managed by Cluster API, we should get this documented in the contract, and implemented also in CAPBK, KCP & CAPD too

Should I align the docs in this same PR? Or would this be a follow up?
I am totally fine with bringing this change to the other providers once this is merged.

File: api/v1alpha4/common_types.go
Comment (vincepri) [inline api/v1alpha4/common_types.go:60]: Should we turn this into a prefix rather than having watches based on key + values?
```suggestion
	WatchLabelPrefix = "cluster.x-k8s.io/watch"
```
It could be used as a label like: "cluster.x-k8s.io/watch/test"

**Review by vincepri — COMMENTED** [review]
Only concern that came up during review is that now the predicate has to be added to each controller separately (which is also hard to test and could break any time).

Is Namespace based filtering not enough to run multiple controllers?

With Namespace based filters we also get the benefit of having the informers and all internal cache to only ever watch that namespace. With label based predicates instead, the cache is still being populated entirely, and watch notifications are still coming through but discarded

File: api/v1alpha4/common_types.go
Comment (MarcelMue) [reply api/v1alpha4/common_types.go:60]: I would be fine doing it either way. Key + value seems a little bit more intuitive to me but I am open to change it.

**Comment by MarcelMue** [comment]
> Is Namespace based filtering not enough to run multiple controllers?

To me this mostly boils down to usability. I agree that there are some performance downside but this is a completely optional flag and from my experience with using this approach, the performance impact is negligible.

The usability is important to me because this enables switching the reconciling controller in a single command: Changing the label. I feel like that is easy to understand, execute and handle.

With a namespaced approach we would firstly have our namespace structure to fit the controller versions we use - this is not very nice IMO.
Additionally switching the version now becomes a rather complex action:
1. Pause all CRs of the Cluster.
2. Move all CRs of the Cluster to a different namespace.
3. Unpause all CRs of the Cluster.

A basically atomic action in the workflow I propose here becomes a rather error prone, multi step process.

I feel like we discussed these points before in a meeting - I am also more than happy to provide E2E tests for this functionality (I think the e2e workthrough is not out yet) if this increases your confidence.

**Comment by MarcelMue** [comment]
@CecileRobertMichon You also had opinions on this IIRC.

File: api/v1alpha4/common_types.go
Comment (CecileRobertMichon) [reply api/v1alpha4/common_types.go:60]: what would the UX look like if using a prefix + suffix? You'd still have to pass in the suffix in the controller, and then you need to make sure you're setting the label the right way, ie. `cluster.x-k8s.io/watch/test` and not `cluster.x-k8s.io/watch-test` or `cluster.x-k8s.io/watch_test`.

IMO, key + value is more intuitive as @MarcelMue said.

**Comment by CecileRobertMichon** [comment]
> Should I align the docs in this same PR? Or would this be a follow up?
I am totally fine with bringing this change to the other providers once this is merged.

Yes, please add to the bootstrap and infra provider contracts in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/bootstrap.md, https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/cluster-infrastructure.md, and https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/machine-infrastructure.md

let's also add a quick note about it in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md (see https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha2-to-v1alpha3.md#support-the-clusterx-k8siopaused-annotation-and-clusterspecpaused-field for an example)

**Comment by CecileRobertMichon** [comment]
> Only concern that came up during review is that now the predicate has to be added to each controller separately (which is also hard to test and could break any time).

I'm +1 on this PR. There is precedent for this with the Pause annotation. In addition, the default behavior does not change and this is completely opt-in, so the risk is low. We should definitely consider adding an E2E test for preventing future regressions though.

**Comment by vincepri** [comment]
> I'm +1 on this PR. There is precedent for this with the Pause annotation

@CecileRobertMichon Did you meant the feature gates, or did I miss the use case you were referring to?

**Comment by CecileRobertMichon** [comment]
@vincepri no I meant the pause annotation predicate (which had to be added to every single controller, including in infra providers)

**Comment by vincepri** [comment]
@CecileRobertMichon Oh I see, the implementation bits, yes. The flag though also has to be kept in sync, which is what I was thinking w.r.t. the feature gates.

Thanks @MarcelMue for the explanation above, as long as we document these requirements and limitations +1

**Comment by CecileRobertMichon** [comment]
@MarcelMue please fix the conflict

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Comment by MarcelMue** [comment]
> @MarcelMue please fix the conflict

Should be all fixed now :+1:

**Comment by MarcelMue** [comment]
@CecileRobertMichon Can you take another look? I know reviewers have been a bit swamped recently :heart:

**Comment by fabriziopandini** [comment]
change SGTM
I would suggest to title the PR with ⚠️ given we are changing the contract and to ask for a quick +1 to provider's maintainer as well.

**Comment by MarcelMue** [comment]
> change SGTM
> I would suggest to title the PR with warning given we are changing the contract and to ask for a quick +1 to provider's maintainer as well.

Done - not sure why it's not rendered as :warning: though :sweat_smile:

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by CecileRobertMichon** [comment]
> I would suggest to title the PR with warning given we are changing the contract and to ask for a quick +1 to provider's maintainer as well.

Azure provider maintainer here - /lgtm

/assign @randomvariable @yastij @cpanato

**Comment by CecileRobertMichon** [comment]
Actually, I just realized we should also update https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/architecture/controllers/support-multiple-instances.md @MarcelMue

**Comment by MarcelMue** [comment]
> Actually, I just realized we should also update https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/architecture/controllers/support-multiple-instances.md @MarcelMue

This PR already contains a simple addition of the label in the same section where the `namespace` flag is mentioned. Would you like to something additional there? I felt like the general statement in that file remained unchanged:
https://github.com/kubernetes-sigs/cluster-api/pull/4119/files#diff-e022c3c4584b8bb87c39431d43bdac3c6d62743735ede072b8a1a1bf9e1d64ce

**Comment by CecileRobertMichon** [comment]
> This PR already contains a simple addition of the label in the same section where the namespace flag is mentioned. Would you like to something additional there

Ahh I must have missed that on my re-read. Thanks, all good from my side then.

**Review by yastij — COMMENTED** [review]
/lgtm

pending rebase, fyi @gab-satchi

**Comment by MarcelMue** [comment]
> /lgtm
> 
> pending rebase, fyi @gab-satchi

Should be all rebased correctly now :)

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Review by vincepri — APPROVED** [review]
/approve

**Comment by vincepri** [comment]
/retitle :warning: Add watch label to allow multiple instances of controllers

**Comment by vincepri** [comment]
/retitle ⚠️ Add watch label to allow multiple manager instances

**Pull request merged**
